File: frontend/src/services/paymentService.ts
import api from './api';
import { Booking } from './bookingService';

// Payment interfaces
export interface PaymentOrder {
  orderId: string;
  paymentSessionId: string;
  amount: number;
  currency: string;
  paymentId: string;
  environment: 'sandbox' | 'production';
  tempBookingData?: Record<string, unknown>;
  isTemporaryBooking?: boolean;
}

export interface PaymentVerification {
  orderId: string;
  bookingId?: string;
}

export interface PaymentData {
  _id: string;
  userId: string;
  bookingId: string;
  orderId: string;
  paymentId?: string;
  amount: number;
  currency: string;
  status: 'created' | 'attempted' | 'paid' | 'failed' | 'refunded' | 'partial_refund';
  cashfreeOrderId?: string;
  cashfreePaymentId?: string;
  failureReason?: string;
  refundId?: string;
  refundAmount: number;
  createdAt: string;
  updatedAt: string;
  user?: {
    name: string;
    email: string;
    phone: string;
  };
  booking?: Booking;
  // Legacy fields for backward compatibility
  platformFee?: number;
  merchantEarnings?: number;
  date?: string;
  billing?: {
    invoiceNumber: string;
    invoiceDate: string;
    fileUrl: string;
    total: number;
  };
}

export interface PaymentHistory {
  payments: PaymentData[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    pages: number;
  };
}

export const paymentService = {
  /**
   * Create Cashfree order
   */
  createOrder: async (bookingId?: string, amount?: number, currency = 'INR', tempBookingData?: Record<string, unknown>): Promise<PaymentOrder> => {
    const requestData: Record<string, unknown> = { 
      amount,
      currency
    };

    if (tempBookingData) {
      requestData.tempBookingData = tempBookingData;
    }
    
    // Only add bookingId if it's a non-empty string and not the string 'undefined'
    if (bookingId && typeof bookingId === 'string' && bookingId.trim() !== '' && bookingId !== 'undefined' && bookingId !== 'null') {
      requestData.bookingId = bookingId;
    }

    console.log('paymentService.createOrder request data:', requestData);
    const response = await api.post('/payments/create-order', requestData);
    console.log('paymentService.createOrder response data:', response.data);
    return response.data.data;
  },

  /**
   * Verify payment with Cashfree
   */
  verifyPayment: async (data: PaymentVerification) => {
    const response = await api.post('/payments/verify', data);
    return response.data;
  },

  /**
   * Get payment by ID
   */
  getPaymentStatus: async (paymentId: string): Promise<PaymentData> => {
    const response = await api.get(`/payments/status/${paymentId}`);
    return response.data.data;
  },

  /**
   * Get user payment history
   */
  getPaymentHistory: async (page = 1, limit = 10): Promise<PaymentHistory> => {
    const response = await api.get(`/payments/history?page=${page}&limit=${limit}`);
    return response.data;
  },

  /**
   * Get all payments (Admin only)
   */
  getAllPayments: async (page = 1, limit = 20, filters?: {
    status?: string;
    startDate?: string;
    endDate?: string;
  }) => {
    const normalizeDate = (dateStr?: string) => {
      if (!dateStr) return undefined;
      const trimmed = dateStr.trim().replace(/\//g, '-');
      
      // If already YYYY-MM-DD
      if (/^\d{4}-\d{2}-\d{2}$/.test(trimmed)) {
        return trimmed;
      }
      
      // If DD-MM-YYYY or MM-DD-YYYY
      if (/^\d{2}-\d{2}-\d{4}$/.test(trimmed)) {
        const parts = trimmed.split('-');
        const year = parts[2];
        const p1 = parts[0];
        const p2 = parts[1];
        
        const y = Number(year);
        const d_ddmmyyyy = Number(p1);
        const m_ddmmyyyy = Number(p2);
        const dateA = new Date(y, m_ddmmyyyy - 1, d_ddmmyyyy);
        const validA = dateA.getFullYear() === y && (dateA.getMonth() + 1) === m_ddmmyyyy && dateA.getDate() === d_ddmmyyyy;
        
        if (validA) {
          return `${year}-${p2.padStart(2, '0')}-${p1.padStart(2, '0')}`;
        }
        
        const m_mmddyyyy = Number(p1);
        const d_mmddyyyy = Number(p2);
        const dateB = new Date(y, m_mmddyyyy - 1, d_mmddyyyy);
        const validB = dateB.getFullYear() === y && (dateB.getMonth() + 1) === m_mmddyyyy && dateB.getDate() === d_mmddyyyy;
        
        if (validB) {
          return `${year}-${p1.padStart(2, '0')}-${p2.padStart(2, '0')}`;
        }
      }
      return dateStr;
    };

    const start = normalizeDate(filters?.startDate);
    const end = normalizeDate(filters?.endDate);

    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
      ...(filters?.status && { status: filters.status }),
      ...(start && { startDate: start }),
      ...(end && { endDate: end })
    });

    const response = await api.get(`/payments/all?${params}`);
    return response.data;
  },

  /**
   * Process refund (Admin only)
   */
  processRefund: async (paymentId: string, amount?: number, reason?: string) => {
    const response = await api.post('/payments/refund', {
      paymentId,
      amount,
      reason
    });
    return response.data;
  },

  retryPayment: async (orderId: string): Promise<PaymentOrder> => {
    const response = await api.post('/payments/retry', { orderId });
    return response.data.data;
  },

  getUserOrders: async () => {
    const response = await api.get('/payments/orders');
    return response.data.data;
  }
};

File: mobile/lib/services/payment_service.dart
import '../core/api_client.dart';
import '../core/env.dart';

class PaymentData {
  final String id;
  final String bookingId;
  final num amount;
  final num platformFee;
  final num merchantEarnings;
  final String status;
  final String? paymentId;
  final String date;
  final String? userName;
  final String? userEmail;

  PaymentData({
    required this.id,
    required this.bookingId,
    required this.amount,
    required this.platformFee,
    required this.merchantEarnings,
    required this.status,
    this.paymentId,
    required this.date,
    this.userName,
    this.userEmail,
  });

  factory PaymentData.fromJson(Map<String, dynamic> json) {
    final user = json['user'];
    return PaymentData(
      id: (json['_id'] ?? '').toString(),
      bookingId: (json['bookingId'] ?? '').toString(),
      amount: (json['amount'] ?? 0) as num,
      platformFee: (json['platformFee'] ?? 0) as num,
      merchantEarnings: (json['merchantEarnings'] ?? 0) as num,
      status: (json['status'] ?? '').toString(),
      paymentId: json['paymentId']?.toString(),
      date: (json['date'] ?? '').toString(),
      userName: user is Map ? user['name']?.toString() : null,
      userEmail: user is Map ? user['email']?.toString() : null,
    );
  }
}

class PaymentService {
  final ApiClient _api = ApiClient();

  // Dummy payment (replaces Razorpay)
  Future<Map<String, dynamic>> processDummyPayment(
    String bookingId, {
    Map<String, dynamic>? tempBookingData,
  }) async {
    return await _api.postJson(
      ApiEndpoints.paymentsDummyPay,
      body: {
        'bookingId': tempBookingData != null ? null : bookingId,
        'tempBookingData': ?tempBookingData,
      },
    );
  }

  // Legacy Razorpay endpoints
  Future<Map<String, dynamic>> createOrder(String bookingId) async {
    return await _api.postJson(
      ApiEndpoints.paymentsCreateOrder,
      body: {'bookingId': bookingId},
    );
  }

  Future<Map<String, dynamic>> verifyPayment(Map<String, dynamic> data) async {
    return await _api.postJson(ApiEndpoints.paymentsVerifyPayment, body: data);
  }

  Future<List<PaymentData>> getAllPayments() async {
    final res = await _api.getAny('/payments');
    final items = <PaymentData>[];
    if (res is List) {
      for (final e in res) {
        if (e is Map<String, dynamic>) {
          items.add(PaymentData.fromJson(e));
        } else if (e is Map) {
          items.add(PaymentData.fromJson(Map<String, dynamic>.from(e)));
        }
      }
    }
    return items;
  }
}
